File: 278. First Bad Version/solution.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        
        if (n <= 0 || !isBadVersion(n)) {
            return -1;
        }
        
        if (isBadVersion(1)) {
            return 1;
        }
        
        int left = 1;
        int right = n;
        
        while(left<right){
            int middle = left + (right-left)/2;
            if(isBadVersion(middle)==false){
                left = middle + 1;
            }else{
                right = middle;
            }
        }
        return left;
    }
}

File: 278. First Bad Version/solution.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

//1,2,3,4,5
//g,g,g,b,b
class Solution {
public:
    int firstBadVersion(int n) {
        int low = 1;
        int high = n;
        
        while(low < high){
            int mid = low + (high-low)/2;
            if(!isBadVersion(mid)){
                low = mid + 1;
            }else{
                high = mid;
            }
        }
        return low;
    }
};
